<?xml version="1.0" ?>

<robot name="my_mm_robot" xmlns:xacro="https://www.ros.org/wiki/xacro" >

  <material name="black">
    <color rgba="0.0 0.0 0.0 1.0"/>
  </material>
  <material name="blue">
    <color rgba="0.203125 0.23828125 0.28515625 1.0"/>
  </material>
  <material name="green">
    <color rgba="0.0 0.8 0.0 1.0"/>
  </material>
  <material name="grey">
    <color rgba="0.2 0.2 0.2 1.0"/>
  </material>
  <material name="orange">
    <color rgba="1.0 0.423529411765 0.0392156862745 1.0"/>
  </material>
  <material name="brown">
    <color rgba="0.870588235294 0.811764705882 0.764705882353 1.0"/>
  </material>
  <material name="red">
    <color rgba="0.80078125 0.12890625 0.1328125 1.0"/>
  </material>
  <material name="white">
    <color rgba="1.0 1.0 1.0 1.0"/>
  </material>

  <gazebo reference="link_chassis">
    <material>Gazebo/Red</material>
  </gazebo>
  <gazebo reference="link_left_rear_wheel">
    <material>Gazebo/Black</material>
  </gazebo>
  <gazebo reference="link_right_front_wheel">
    <material>Gazebo/Black</material>
  </gazebo>
    <gazebo reference="link_left_front_wheel">
    <material>Gazebo/Black</material>
  </gazebo>
  <gazebo reference="link_right_rear_wheel">
    <material>Gazebo/Black</material>
  </gazebo>


  <link name="link_chassis">
    <!-- pose and inertial -->
    <pose>0 0 0.1 0 0 0</pose>

    <inertial>
      <mass value="0.5"/>
      <origin rpy="0  0  0" xyz="0 0.01 0"/>
      <inertia ixx="0.00022083" ixy="0" ixz="0" iyy="0.00022083" iyz="0" izz="0.00040833"/> 
    </inertial>

    <collision name="link_chassis_collision">
      <origin rpy="0  0  0" xyz="0     0    -0.005"/>
      <geometry>
        <box size="0.07 0.07 0.02"/>
      </geometry>
    </collision>

    <visual name="link_chassis_visual">
      <origin rpy="0  0  0" xyz="0     0    -0.005"/>
      <geometry>
        <box size="0.07 0.07 0.02"/>
      </geometry>
    </visual>


</link>

  <!-- Create wheel right -->

  <link name="link_right_front_wheel">
    <inertial>
      <mass value="0.2"/>
      <origin rpy="0 1.5707 0" xyz="0 0 0"/>
      <inertia ixx="1.17e-06" ixy="0" ixz="0" iyy="1.17e-06" iyz="0" izz="2.25e-06"/>
    </inertial>

    <collision name="link_right_front_wheel_collision">
      <origin rpy="0 1.5707 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.005" radius="0.015"/>
      </geometry>
    </collision>

    <visual name="link_right_front_wheel_visual">
      <origin rpy="0 1.5707 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.005" radius="0.015"/>
      </geometry>
    </visual>

  </link>

  <!--   Joint for right wheel  -->
  <joint name="joint_right_front_wheel" type="continuous">
    <origin rpy="0 0 0" xyz="-0.04  -0.03 -0.015"/>
    <child link="link_right_front_wheel" />
    <parent link="link_chassis"/>
    <axis rpy="0 0 0" xyz="1 0 0"/>
    <limit effort="5" velocity="6"/>
    <joint_properties damping="10000.0" friction="10000.0" />
  </joint>
  <transmission name="right_front_wheel_trans">
     <type>transmission_interface/SimpleTransmission</type>
     <actuator name="right_front_wheel_motor">
       <hardwareInterface>VelocityJointInterface</hardwareInterface>
       <mechanicalReduction>7</mechanicalReduction>
     </actuator>
     <joint name="joint_right_front_wheel">
       <hardwareInterface>VelocityJointInterface</hardwareInterface>
     </joint>
  </transmission>
  
   <link name="link_right_rear_wheel">
    <inertial>
      <mass value="0.2"/>
      <origin rpy="0 1.5707 0" xyz="0 0 0"/>
      <inertia ixx="1.17e-06" ixy="0" ixz="0" iyy="1.17e-06" iyz="0" izz="2.25e-06"/>
    </inertial>

    <collision name="link_right_rear_wheel_collision">
      <origin rpy="0 1.5707 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.005" radius="0.015"/>
      </geometry>
    </collision>

    <visual name="link_right_rear_wheel_visual">
      <origin rpy="0 1.5707 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.005" radius="0.015"/>
      </geometry>
    </visual>

  </link>

  <!--   Joint for right wheel  -->
  <joint name="joint_right_rear_wheel" type="continuous">
    <origin rpy="0 0 0" xyz="-0.04  0.03 -0.015"/>
    <child link="link_right_rear_wheel" />
    <parent link="link_chassis"/>
    <axis rpy="0 0 0" xyz="1 0 0"/>
    <limit effort="5" velocity="6"/>
    <joint_properties damping="10000.0" friction="10000.0" />
  </joint>
  <transmission name="right_rear_wheel_trans">
     <type>transmission_interface/SimpleTransmission</type>
     <actuator name="right_rear_wheel_motor">
       <hardwareInterface>VelocityJointInterface</hardwareInterface>
       <mechanicalReduction>7</mechanicalReduction>
     </actuator>
     <joint name="joint_right_rear_wheel">
       <hardwareInterface>VelocityJointInterface</hardwareInterface>
     </joint>
  </transmission>


  <!-- Left Wheel link -->

  <link name="link_left_rear_wheel">
    <inertial>
      <mass value="0.2"/>
      <origin rpy="0 1.5707 0" xyz="0 0 0"/>
      <inertia ixx="1.17e-06" ixy="0" ixz="0" iyy="1.17e-06" iyz="0" izz="2.25e-06"/>
    </inertial>

    <collision name="link_left_rear_wheel_collision">
      <origin rpy="0 1.5707 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.005" radius="0.015"/>
      </geometry>
    </collision>

    <visual name="link_left_rear_wheel_visual">
      <origin rpy="0 1.5707 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.005" radius="0.015"/>
      </geometry>
    </visual>

  </link>

  <!--   Joint for left_rear wheel  -->
  <joint name="joint_left_rear_wheel" type="continuous">
    <origin rpy="0 0 0" xyz="0.04 0.03 -0.015"/>
    <child link="link_left_rear_wheel" />
    <parent link="link_chassis"/>
    <axis rpy="0 0 0" xyz="1 0 0"/>
    <limit effort="5" velocity="6"/>
    <joint_properties damping="10000.0" friction="10000.0" />
  </joint>
  <transmission name="left_rear_wheel_trans">
     <type>transmission_interface/SimpleTransmission</type>
     <actuator name="left_rear_wheel_motor">
       <hardwareInterface>VelocityJointInterface</hardwareInterface>
       <mechanicalReduction>7</mechanicalReduction>
     </actuator>
     <joint name="joint_left_rear_wheel">
       <hardwareInterface>VelocityJointInterface</hardwareInterface>
     </joint>
  </transmission>
  <!-- Left Wheel link -->

  <link name="link_left_front_wheel">
    <inertial>
      <mass value="0.2"/>
      <origin rpy="0 1.5707 0" xyz="0 0 0"/>
      <inertia ixx="1.17e-06" ixy="0" ixz="0" iyy="1.17e-06" iyz="0" izz="2.25e-06"/>
    </inertial>

    <collision name="link_left_front_wheel_collision">
      <origin rpy="0 1.5707 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.005" radius="0.015"/>
      </geometry>
    </collision>

    <visual name="link_left_front_wheel_visual">
      <origin rpy="0 1.5707 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.005" radius="0.015"/>
      </geometry>
    </visual>

  </link>

  <!--   Joint for left_font wheel  -->
  <joint name="joint_left_front_wheel" type="continuous">
    <origin rpy="0 0 0" xyz="0.04 -0.03 -0.015"/>
    <child link="link_left_front_wheel" />
    <parent link="link_chassis"/>
    <axis rpy="0 0 0" xyz="1 0 0"/>
    <limit effort="5" velocity="6"/>
    <joint_properties damping="10000.0" friction="10000.0" />
  </joint>
  <transmission name="left_front_wheel_trans">
     <type>transmission_interface/SimpleTransmission</type>
     <actuator name="left_front_wheel_motor">
       <hardwareInterface>VelocityJointInterface</hardwareInterface>
       <mechanicalReduction>7</mechanicalReduction>
     </actuator>
     <joint name="joint_left_front_wheel">
       <hardwareInterface>VelocityJointInterface</hardwareInterface>
     </joint>
  </transmission>

  <gazebo>
 
  <plugin name="skid_steer_drive_controller" filename="libgazebo_ros_skid_steer_drive.so">
    <updateRate>20.0</updateRate>
    <robotNamespace>/</robotNamespace>
    <leftFrontJoint>joint_left_front_wheel</leftFrontJoint>
    <rightFrontJoint>joint_right_front_wheel</rightFrontJoint>
    <leftRearJoint>joint_left_rear_wheel</leftRearJoint>
    <rightRearJoint>joint_right_rear_wheel</rightRearJoint>
    <wheelSeparation>0.4</wheelSeparation>
    <wheelDiameter>0.2</wheelDiameter>
    <commandTopic>cmd_vel</commandTopic>
    <robotBaseFrame>link_chassis</robotBaseFrame>
    <torque>20</torque>
    <odometryTopic>odom</odometryTopic>
    <odometryFrame>odom</odometryFrame>
    <topicName>cmd_vel</topicName>
    <broadcastTF>1</broadcastTF>
    <covariance_x>0</covariance_x>
    <covariance_y>0</covariance_y>
    <covariance_yaw>0</covariance_yaw>
  </plugin>


</gazebo>



    <!--
    Adding a laser scan sensor to the robot
  -->
  <link name="sensor_laser">
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <mass value="0.01" />
      <inertia ixx="0.02" ixy="0" ixz="0" iyy="0.02" iyz="0" izz="0.02"/>
    </inertial>

    <visual>
      <origin xyz="0.02 0 0" rpy="0 0 0" />
      <geometry>
        <cylinder radius="0.005" length="0.02"/>
      </geometry>
      <material name="white" />
    </visual>

    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.005" length="0.001"/>
      </geometry>
    </collision>
  </link>

  <joint name="joint_sensor_laser" type="fixed">
    <origin rpy="0  0  -1.57" xyz="0    0    0.01"/>
    <parent link="link_chassis"/>
    <child link="sensor_laser"/>
  </joint>


  <gazebo reference="sensor_laser">
    <sensor type="ray" name="head_hokuyo_sensor">
      <pose>0 0 0 0 0 0</pose>
      <visualize>true</visualize>
      <update_rate>20</update_rate>
      <ray>
        <scan>
          <horizontal>
            <samples>720</samples>
            <resolution>1</resolution>
            <min_angle>-1.570796</min_angle>
            <max_angle>1.570796</max_angle>
          </horizontal>
        </scan>
        <range>
          <min>0.04</min>
          <max>10.0</max>
          <resolution>0.01</resolution>
        </range>
      </ray>
      <plugin name="gazebo_ros_head_hokuyo_controller" filename="libgazebo_ros_laser.so">
        <topicName>/my_mm_robot/laser/scan</topicName>
        <frameName>sensor_laser</frameName>
        <broadcastTF>1</broadcastTF>
      </plugin>
    </sensor>
  </gazebo>

  <!-- <gazebo>
    <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
      <robotNamespace>/simple_model</robotNamespace>
      <legacyModeNS>true</legacyModeNS>
      <robotSimType>gazebo_ros_control/DefaultRobotHWSim</robotSimType>
    </plugin>
  </gazebo>-->

</robot>
